fairscale/experimental/nn/offload.py

```diff
@@ -188,16 +188,19 @@ def forward(ctx: Any, inputs: Any, dummy_input: Any, model_instance: Any) -> Any
 
             output = output if isinstance(output, tuple) else (output,)
             with torch.autograd.profiler.record_function("fairscale.experimental.nn.offload:forward_drop"):
-                # The last instance will lose the gradient function if we move it to the CPU.
-                # This is because all grad function are present on the device that ran the FW pass.
-                if index == len(model_instance.model_slices) - 1:
-                    model_instance._activations.append(output)
-                else:
-                    model_instance._activations.append(tuple([a.cpu() for a in list(output)]))
+                # Move the activation used back for the curent shard back to the CPU.
+                model_instance._activations[index] = tuple([a.cpu() for a in list(model_instance._activations[index])])
+                # The newly computed activations remain on the GPU ready for the next shard computation.
```

Comment:
it's not obvious to me where they are moved or dropped ?

Reply:
Each intermediate activation gets moved to the CPU once it has been used for the FW pass of a shard. The output of the previous shard is the input of the next shard which is why we don't move the output to the CPU before processing the next shared. Line 192 moves the activation to the CPU since it is done for now. Line 194 represents the new activation which will be on the GPU waiting for the next shard. 

Not the simplest explanation but I'll try to improve the docs.